compute/change_machine_type.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package snippets
+
+// [START compute_instance_machine_type_update]
+import (
+	"context"
+	"fmt"
+	"io"
+
+	compute "cloud.google.com/go/compute/apiv1"
+	computepb "google.golang.org/genproto/googleapis/cloud/compute/v1"
+)
+
+// changeMachineType changes the machine type of the instance.
+// If the instance is running, it will shut it down
+// so that the machine type could be changed.
+func changeMachineType(
+	w io.Writer,
+	projectID, zone, instanceName, newInstanceType string,
+) error {
+	// projectID := "your_project_id"
+	// zone := "europe-central2-b"
+	// instanceName := "your_instance_name"
+	// newInstanceType = "e2-standard-2"
+
+	ctx := context.Background()
+	instancesClient, err := compute.NewInstancesRESTClient(ctx)
+	if err != nil {
+		return fmt.Errorf("NewInstancesRESTClient: %v", err)
+	}
+	defer instancesClient.Close()
+
+	reqInstance := &computepb.GetInstanceRequest{
+		Project:  projectID,
+		Zone:     zone,
+		Instance: instanceName,
+	}
+
+	instance, err := instancesClient.Get(ctx, reqInstance)
+	if err != nil {
+		return fmt.Errorf("unable to get instance: %v", err)
+	}
+
+	containsString := func(s []string, str string) bool {
+		for _, v := range s {
+			if v == str {
+				return true
+			}
+		}
+
+		return false
+	}
+
+	// Make sure that the machine is turned off
+	if !containsString([]string{"TERMINATED", "STOPPED"}, instance.GetStatus()) {
```

Comment:
I think there's an easier way to do this that makes it clearer what you're trying to do.

```
if !strings.Contains("STOPPED TERMINATED", instance.GetStatus()) {
      // ...
}

```

Reply:
Is this a Go idiom? It feels like a clever hack to me, and it took me a while to understand what's actually happening here.

This is copied from another sample too, so it may also be a good idea to keep them the same for consistency.

Reply:
Unfortunately Go doesn't have an easy way to determine whether an element is a member of a slice or array. One way, of course is to write a loop that determines membership (original PR code). Another way is to search for substrings.

My comment is merely a suggestion of an alternative. Feel free to accept or reject as you see fit.

Reply:
I see. I'm going to keep the current implementation for consistency with other samples, then.